### PR TITLE
fix(nginx): fix 404 when refreshing page with custom nginx.conf

### PR DESCRIPTION
**Pull request by doc-code-hub**
Refering to #1683 i've made this simple edit by adding a custom nginx.conf and adding it into the Dockerfile.

I started from the default nginx.conf ([link](https://gist.githubusercontent.com/nishantmodak/d08aae033775cb1a0f8a/raw/ebea0ae66e0a0188009accae2acba88cc646ddcd/nginx.conf.default)) and I modified some parts by adding in particular:
```nginx
 location / {
           # First attempt to serve request as file, then
           # as directory, then fall back to redirecting to index.html
           try_files $uri $uri/ $uri.html /index.html;
        }
```

This requires no action by users to deploy and solve the problem!

### CODE REVIEW

**Comment by ferferga** [comment]
This fixes Docker but leaves users that deploy their client using their own reverse proxy with the need to tweak their own config (which means reading a documentation, etc, and not everybody is willing to do so, when we're aiming to make this as simple to use as possible).

This doesn't also cover the cases for when the client is hosted under a Jellyfin server (like the current web client does), which is another situation [we might need to cover](https://github.com/jellyfin/jellyfin-vue/discussions/1679)

PS: We also used our custom nginx config before, but we removed it because [it was yet another thing to maintain for ourselves](https://github.com/jellyfin/jellyfin-vue/pull/1645)

Please refer to #1684 and, if that fixes your issue, please close this pull request.

**Comment by ThibaultNocchi** [comment]
Hey, so we're going this way:

We'll move to hash by default. But, "history" mode (that's how current routing is named) will be toggleable with an env var. It'll also be used in our Dockerfiles. So we'll first merge ferf PR to set the new default, and we'll continue with you on this PR. :)

**Comment by ThibaultNocchi** [comment]
So, I see a few points:

* You should copy the Nginx conf to `/etc/nginx/conf.d/nginx.conf` so you can have an easier config
* That easier config can drop everything except the `server` block. In that one you can simplify it to:

```
    server {
        # You would want to make a separate file with its own server block for each virtual domain
        # on your server and then include them.
        listen       80;
        #tells Nginx the hostname and the TCP port where it should listen for HTTP connections.
        # listen 80; is equivalent to listen *:80;

        root /usr/share/nginx/html;

        location / {
           # First attempt to serve request as file, then
           # as directory, then fall back to redirecting to index.html
           try_files $uri $uri/ /index.html;
        }
    }
```

So we dropped the server_name as that *should* work without and edited the try_files to match VueRouter config file https://router.vuejs.org/guide/essentials/history-mode.html#nginx

Tell me if you need help figuring out how to enable "history" mode with env vars. :)

**Comment by doc-code-hub** [comment]
Fantastic, I like the possibility of being able to choose what you propose!

Later I will try to implement the change (and the nginx.conf update as you suggest)! (around 13:00 Italian time)

If I have trouble figuring it out, I'll definitely come back here to ask for some help :)

Thank you!

**Comment by ferferga** [comment]
@doc-code-hub Btw in case you aren't yet signed up in Matrix, present yourself in our [Vue development room](https://matrix.to/#/#jellyfin-vue:matrix.org) (we also have it bridged to Discord) so we can discuss this kind of stuff in a more informal way!

**Comment by doc-code-hub** [comment]
Very gladly :)

Later I will definitely do it, thanks!

File: Dockerfile [hunk 1]
```diff
@@ -20,15 +26,35 @@ RUN npm ci --no-audit
...
 FROM nginx:alpine
+
+# Set environment variables
+ARG HISTORY_ROUTER_MODE=false
+
 COPY --from=build /app/src/dist/ /usr/share/nginx/html/
+
+COPY --from=build /app/.docker/nginx.conf /etc/nginx/conf.d/custom.conf 
+
+# If specified, set's custom nginx.conf to support spa (for 'history' routing mode)
```
Comment (ferferga) [inline Dockerfile:44]: I would honestly remove all this block and just copy our custom config no matter the history mode. The default config gives the same end result to the user (the only difference is the try block) and this just clutters our Dockerfile for the little benefit it provides 

*(now at least, maybe in the future it's useful for E2E testing)*

Thoughts @ThibaultNocchi?

Comment (doc-code-hub) [reply Dockerfile:44]: I was thinking that a custom nginx.conf whoud be a problem for some cases [(your comment)](https://github.com/jellyfin/jellyfin-vue/pull/1685#issuecomment-1009509281)

No?

Comment (ferferga) [reply Dockerfile:44]: Oh yes, but I meant users that dont use Docker (not everyone does).  
If an user builds the client himself or pulls it as a deb (like current Web) or from a release (we don't have debs or releases, but once we reach an stable state we will) and serves the client through his webserver of choice, the user must configure it with the ``try_files`` directive.


Docker in that regard its our walled garden and we can provide our best config where everything is beautiful. But we can't expect everybody will have our ideal configuration, hence the need for the default to be hash. 

In this repo we are "selling" users two "bundles":

* The client (our "best-seller", as its what the repo is about) 

* Client + webserver (Docker) 

This repo is all about the client and, as is, it should work as universally as possible with the default config. History mode doesn't allow that as it adds the user's webserver into the mix, and adds variables out of our control. An user with a misconfigured webserver will have a degraded experience, so we should stick to "the universal best". 

However, we're offering too (through Docker) the mix client + webserver: with it, we remove the webserver variable from the user's hands and bring it to us. This means that for users choosing our bundle, we can give them what we think are the best defaults (we all agree is history mode). 

To sum up: the file does not hurt and provides "universally" safe defaults. I also expect nobody in Docker will use the hash mode, because it makes no sense for them (most people just fire docker build or pull the image, and those will use history mode by default). 
I only expect people building outside Docker to care about changing that option, since those are the ones who are likely not going to use "our walled garden" and are more likely to be scatterbrained about their webserver config. 

That's why the current config is ideal: outside docker we use hash, inside it we use history. The environments where they both make more sense.

Comment (doc-code-hub) [reply Dockerfile:44]: Then in this case it is better to change the default ARG to true, 
to pass the --build-arg only in the unlikely event that you want to disable. Do you agree?

bc now default is 'hash' for docker too

File: Dockerfile [hunk 2]
```diff
@@ -20,15 +26,22 @@ RUN npm ci --no-audit
...
+
+COPY --from=build /app/.docker/nginx.conf /etc/nginx/conf.d/default.conf 
+
 EXPOSE 80
 
 # Set labels
 LABEL maintainer="Jellyfin Packaging Team - packaging@jellyfin.org"
 LABEL org.opencontainers.image.source="https://github.com/jellyfin/jellyfin-vue"
 LABEL org.opencontainers.image.description "Commit: ${NUXT_ENV_COMMIT}"
+LABEL org.opencontainers.image.description "History Router Mode: ${HISTORY_ROUTER_MODE}"
```
Comment (ferferga) [inline Dockerfile:47]: Doing duplicates remove the previous one: https://docs.docker.com/config/labels-custom-metadata/#label-keys-and-values

```suggestion
LABEL org.opencontainers.image.description="Commit: ${NUXT_ENV_COMMIT} History router mode: ${HISTORY_ROUTER_MODE}"
```

File: .docker/nginx.conf [hunk 3]
```diff
@@ -0,0 +1,15 @@
+    server {
+        # You would want to make a separate file with its own server block for each virtual domain
+        # on your server and then include them.
+        listen       80;
```
Comment (ferferga) [inline .docker/nginx.conf:4]:
```suggestion
        listen 80;
```

File: .docker/nginx.conf [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        #tells Nginx the hostname and the TCP port where it should listen for HTTP connections.
+        # listen 80; is equivalent to listen *:80;
+
+        root /usr/share/nginx/html;
+
+        location / {
+           # First attempt to serve request as file, then
+           # as directory, then fall back to redirecting to index.html
```
Comment (ferferga) [inline .docker/nginx.conf:12]:
```suggestion
           # First attempt to serve request as file, then
           # as directory, then fall back to redirecting to index.html
           # This is needed for history mode in Vue router
```

File: .docker/nginx.conf [hunk 5]
```diff
@@ -0,0 +1,15 @@
+    server {
+        # You would want to make a separate file with its own server block for each virtual domain
+        # on your server and then include them.
+        listen       80;
+        #tells Nginx the hostname and the TCP port where it should listen for HTTP connections.
+        # listen 80; is equivalent to listen *:80;
```
Comment (ferferga) [inline .docker/nginx.conf:6]: Not relevant for us

```suggestion
```

File: .docker/nginx.conf [hunk 6]
```diff
@@ -0,0 +1,15 @@
+    server {
+        # You would want to make a separate file with its own server block for each virtual domain
+        # on your server and then include them.
```
Comment (ferferga) [inline .docker/nginx.conf:3]: Not relevant for us

```suggestion
```

File: Dockerfile [hunk 7]
```diff
@@ -1,8 +1,14 @@
...
+
 FROM node:16-alpine AS build
+
 # Set environment variables
+ARG HISTORY_ROUTER_MODE=true
+ENV HISTORY_ROUTER_MODE=$HISTORY_ROUTER_MODE
+
 ARG IS_STABLE=0
+
 ENV NUXT_ENV_COMMIT=""
```
Comment (ferferga) [inline Dockerfile:12]: Keep them together and remove the excess of spaces:

```suggestion
FROM node:16-alpine AS build

# Set build arguments
ARG HISTORY_ROUTER_MODE=true
ARG IS_STABLE=0

# Set environment variables
ENV HISTORY_ROUTER_MODE=$HISTORY_ROUTER_MODE
ENV NUXT_ENV_COMMIT=""
```

File: Dockerfile [hunk 8]
```diff
@@ -20,15 +26,22 @@ RUN npm ci --no-audit
 # Set commit hash
 RUN if [[ $IS_STABLE == "0" ]] ; then NUXT_ENV_COMMIT=$(git rev-parse HEAD) ; fi
 
+
```
Comment (ferferga) [inline Dockerfile:29]:
```suggestion
```

File: Dockerfile [hunk 9]
```diff
@@ -20,15 +26,22 @@ RUN npm ci --no-audit
...
 # Build client
 RUN npm run build
 
+
```
Comment (ferferga) [inline Dockerfile:33]:
```suggestion
```

File: Dockerfile [hunk 10]
```diff
@@ -20,15 +26,22 @@ RUN npm ci --no-audit
...
 # Deploy built distribution to nginx
 FROM nginx:alpine
+
 COPY --from=build /app/src/dist/ /usr/share/nginx/html/
+
```
Comment (ferferga) [inline Dockerfile:38]:
```suggestion
```

**Review by ferferga — CHANGES_REQUESTED** [review]
Some "linting" stuff for our Dockerfile and then LGTM

File: Dockerfile (hunk 1 above)
Comment (ThibaultNocchi) [reply Dockerfile]: I agree that we can copy the config in any case, it's basic SPA routing anyway. :)

**Comment by ThibaultNocchi** [comment]
I'd like to add something: for the env var, drop the true or false value and keep it as bash scripts usually do: either you put fill the env var to enable a feature `HISTORY_ROUTER_MODE=1` or you don't fill it and so the default is used. That'd also simplify the JS file to just detect the presence of the env var.

**Review by ferferga — APPROVED** [review]

**Review by ferferga — APPROVED** [review]

**Comment by doc-code-hub** [comment]
@ThibaultNocchi `That'd also simplify the JS file to just detect the presence of the env var.`

Just for the record:
I tried to limit myself to checking the presence of the variable ( 
```ts
provess.env.HISTORY_ROUTER_MODE ? 'history' : 'hash'
```
), but I still had to keep the check 
```ts
provess.env.HISTORY_ROUTER_MODE === '1'
```
 because the environment variable arrives as a string, for this reason, both '1' and '0' were true ( using history mode).

Just to let you know

**Review by ThibaultNocchi — APPROVED** [review]
